src/main/java/io/epirus/console/deploy/DeployRunner.java

```diff
@@ -0,0 +1,158 @@
+                    .deploy();
+        } else {
+            Console.exitError(USAGE);
+        }
+    }
+
+    public DeployRunner(Network network, AccountManager accountManager, Path workingDirectory) {
+        this.workingDirectory = workingDirectory;
+        this.network = network;
+        Path walletPath = ProjectUtils.loadProjectWalletFile(workingDirectory).get();
+        Path walletPasswordPath =
+                ProjectUtils.loadProjectPasswordWalletFile(workingDirectory).get();
+        this.credentials = ProjectUtils.createCredentials(walletPath, walletPasswordPath);
+        this.accountManager = accountManager;
+    }
+
+    public DeployRunner(Network network, AccountManager accountManager) {
+        this.workingDirectory = Paths.get(System.getProperty("user.dir"));
+        this.network = network;
+        Path walletPath = ProjectUtils.loadProjectWalletFile(workingDirectory).get();
+        Path walletPasswordPath =
+                ProjectUtils.loadProjectPasswordWalletFile(workingDirectory).get();
+        this.credentials = ProjectUtils.createCredentials(walletPath, walletPasswordPath);
+        this.accountManager = accountManager;
+    }
+
+    public void deploy() throws Exception {
+        AccountUtils.accountInit(accountManager);
+        this.accountManager.checkIfAccountIsConfirmed();
+        fundWallet();
+        waitForBalanceUpdate();
+        uploadSolidityMetadata(network, workingDirectory);
+        runGradle(workingDirectory);
+    }
+
+    private void fundWallet() {
+        int tries = 5;
+        while (tries-- > 0) {
+            try {
+                BigInteger accountBalance = accountManager.getAccountBalance(credentials, network);
+                if (accountBalance.equals(BigInteger.ZERO)) {
+                    WalletFunder.fundWallet(
+                            credentials.getAddress(),
+                            Faucet.valueOf(network.getNetworkName().toUpperCase()),
+                            "alex");
+                }
+                return;
+            } catch (Exception e) {
+
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException ex) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+    }
+
+    private void waitForBalanceUpdate() {
+        System.out.println("Checking the account balance...");
+        int tries = 5;
+        while (tries-- > 0) {
+            try {
+                BigInteger accountBalance = accountManager.getAccountBalance(credentials, network);
+                if (!accountBalance.equals(BigInteger.ZERO)) {
+                    return;
+                } else {
+                    Thread.sleep(10000);
+                }
+            } catch (Exception e) {
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException ex) {
+                    Thread.currentThread().interrupt();
```

Comment:
It might actually be nice to use an exponential backoff library as an external dependency if we're going to be doing this sort of thing a lot - makes things much more readable & reduces nesting.

Reply:
this is a valid point but from what from a quick search okhttp does not provide that kind of functionallity. One example https://github.com/square/okhttp/issues/2489.